// @uniswap/universal-router/typechain/factories/IProtocolFeeController__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer } from "ethers";
import { Provider } from "@ethersproject/providers";

import type { IProtocolFeeController } from "../IProtocolFeeController";

export class IProtocolFeeController__factory {
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IProtocolFeeController {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as IProtocolFeeController;
  }
}

const _abi = [
  {
    inputs: [
      {
        components: [
          {
            internalType: "Currency",
            name: "currency0",
            type: "address",
          },
          {
            internalType: "Currency",
            name: "currency1",
            type: "address",
          },
          {
            internalType: "uint24",
            name: "fee",
            type: "uint24",
          },
          {
            internalType: "int24",
            name: "tickSpacing",
            type: "int24",
          },
          {
            internalType: "contract IHooks",
            name: "hooks",
            type: "address",
          },
        ],
        internalType: "struct PoolKey",
        name: "poolKey",
        type: "tuple",
      },
    ],
    name: "protocolFeeForPool",
    outputs: [
      {
        internalType: "uint24",
        name: "protocolFee",
        type: "uint24",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];
